### PR TITLE
How to test Panel data apps

### LINKED ISSUE #4243: Document how to test panel
I would like it to be easy for my self and other Panel users to test panel apps. This includes

- unittests with pytest
- performance tests with pytest-benchmark
- ui tests with PlayWright
- Load tests with Locust + PlayWright

(I believe the mentioned frameworks are the best)

Please document this.

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
Closes #4243 

You can preview the *how to guides* here 

- https://github.com/holoviz/panel/blob/feature/document-test/doc/how_to/test/pytest.md
- https://github.com/holoviz/panel/blob/feature/document-test/doc/how_to/test/uitests.md
- https://github.com/holoviz/panel/blob/feature/document-test/doc/how_to/test/loadtests.md

## Issues experienced or identified during testing

- https://github.com/SvenskaSpel/locust-plugins/issues/101#issuecomment-1367216919

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
## Assets

Some assets used in the docs that should later be uploaded to a cdn.

**app.mp4**

https://user-images.githubusercontent.com/42288570/209909098-e3713542-4b76-41a2-a4ce-c5f50fd7a0d4.mp4

**app.gif**

**Comment by MarcSkovMadsen** [comment]
Hi @philippjfr 

Please make a first draft review. I would like some feedback on the overall outline and content. As a minimum I would appreciate feedback on:

- Is it ok to have all this content in one "how to" guide? If not how should it be split?
- Whether the code in this *how to* should somehow be automatically tested to avoid regressions later on?
- What should I do additionally to bring it to a state where is should be possible to get it approved?

**Comment by philippjfr** [comment]
My initial reaction is that this is indeed a lot of content for a single guide. I'd suggest splitting it into at least two guides:

- Testing Panel UI with Playwright
- Load testing with Locust

File: doc/how_to/test/test.md [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+        pn.state._indicators.clear()
+        pn.state._locations.clear()
+        pn.state._templates.clear()
+        pn.state._views.clear()
+        pn.state._loaded.clear()
+        pn.state.cache.clear()
+        pn.state._scheduled.clear()
+        if pn.state._thread_pool is not None:
+            pn.state._thread_pool.shutdown(wait=False)
+            pn.state._thread_pool = None
```
Comment (philippjfr) [inline doc/how_to/test/test.md:219]: We should probably add some public method to do this.

Comment (maximlt) [reply doc/how_to/test/test.md:219]: Cleaning up the state isn't only useful for UI tests, right?

**Review by maximlt — COMMENTED** [review]
Documenting how to test Panel apps is a great addition!

However, I would recommend splitting this into different how-tos. I believe most users should only be concerned with simple unit tests with Pytest. In its current form the how-to can lead users into an overly complicated path, specially beginners and non developers.

File: doc/how_to/test/test.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+    def _load(self):
+        time.sleep(self.load_delay)
+        return "Loaded"
+
+    def _run(self):
+        time.sleep(self.run_delay)
+        return f"Result {self.runs+1}"
+
+if __name__.startswith("bokeh"):
```
[the remark this answers is not in the document]
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: I've used this in a thousand places over the last 3 years. Would be nice to know if there is another prefered, but not documented way of doing this?

File: doc/how_to/test/test.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+![panel-locust.gif](https://user-images.githubusercontent.com/42288570/209923009-521554d4-dcf8-49b3-8cca-c714037af901.gif)
+
+The median response time (on my laptop) is ~530ms when one user requests the page every second.
+
+If you try to increase to 10 simultanous users you will see a median response time of ~5300ms. If this is a likely scenario, you will have to look into how to improve the performance of your app.
+
+According to [locust-plugins](https://github.com/SvenskaSpel/locust-plugins) it should also be possible to possible to combine Locust and Playwright to test more advanced interactions. Unfortunately it does not work for me on Windows. You can check out the issue with code [here](https://github.com/SvenskaSpel/locust-plugins/issues/101#issuecomment-1367216919).
+
+### Test advanced interactions with Loadwright
```
[the remark this answers is not in the document]
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: 😄 

There is no other way currently. And I also need it for other the next how to guides I plan to write on "performance optimization" and "dask".

Real Load Testing of Panel data apps is has really, really been missing. This is why the Bokeh/ Panel server still has performance issues. We have not been able to get insights and provide reproducible examples. This makes it possible. I've specifically pinned the version here. And its not more *alpha* than any new example or code contributed to Panel.

File: doc/how_to/test/test.md [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+pip install panel pytest pytest-benchmark locust-plugins pytest-playwright playwright pytest-async loadwright==0.1.0
+```
+
+and install the browsers
+
+```bash
+playwright install
+```
+
+## Create the app
```
[the remark this answers is not in the document]
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: I sort of agree.

But this app can be used for all the how to guides from testing, over performance to dask that I have been working on. Its developed and chosen with those insights.

File: doc/how_to/test/test.md (hunk 1 above)
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: Its probably useful for any kind of test where you persist some state. For example if you set the `cache`.

File: doc/how_to/test/test.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+test_run_performance     308.4448  311.8413  310.1718  1.2477  310.1379  1.5623       2;0  3.2240       5           1
+---------------------------------------------------------------------------------------------------------------------
+
+Legend:
+  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
+  OPS: Operations Per Second, computed as 1 / Mean
+================================================================== 4 passed in 4.82s
+```
+
+Notice how we used the `benchmark` fixture of [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/) to test the performance of the `run` event.
```
[the remark this answers is not in the document]
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: I've moved the benchmarking to a separate sections.

File: doc/how_to/test/test.md [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+
+## Test the frontend with pytest and Playwright
+
+Sometimes, for example when you create [custom components](https://panel.holoviz.org/user_guide/Custom_Components.html), it can be useful to test the frontend behaviour.
+
+For testing the frontend we recommend the framework [Playwright](https://playwright.dev/) by Microsoft. Panel it self is tested by this framework.
```
[the remark this answers is not in the document]
Comment (MarcSkovMadsen) [reply doc/how_to/test/test.md]: Thanks. I've used this suggestion in the refactored guide.

**Comment by MarcSkovMadsen** [comment]
Thanks for all the good comments. I've refactored. Please make a second review. Thanks.

File: doc/how_to/test/test.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply doc/how_to/test/test.md]: I'm saying alpha as this is the disclaimer on the project's README.

> Please note this project is at a very early stage an the api and functionality will change!

I would personally be hesitating when advised to depend on a project that states this, as it feels like it's still being built and that I'll have to change a lot of my code in the future.

File: doc/how_to/test/test.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (philippjfr) [reply doc/how_to/test/test.md]: I think that's a very good point but can be addressed by writing a new "Write apps that are easy to test" how-to.

**Comment by philippjfr** [comment]
Thanks for your fixes. I've made a number of edits. One thing that would be nice is if you could re-record some of the gifs with a larger font size. You can do that any time and I'll upload the new versions.

**Comment by MarcSkovMadsen** [comment]
@philippjfr . The link https://assets.holoviz.org/panel/gifs/loadwright.gif used in the docs does not exist. Could you please upload the file? Thanks.

**Comment by philippjfr** [comment]
Done

**Comment by MarcSkovMadsen** [comment]
I've increased the fontsize of the `uitest.gif` and `loadwright.gif` files above. Please take a look and upload them if they are ok @philippjfr . Thanks.

**Comment by MarcSkovMadsen** [comment]
I've tested the how to guides and fixed any errors identified. I don't plan to do more unless told what to do :-)

**Pull request merged**